File: DJIStreamDemo/RTMPServerObjc/TVUIRLBandwidthMeter.m
//
//  TVUIRLBandwidthMeter.m
//  DJIStreamDemo
//

#import "TVUIRLBandwidthMeter.h"

@interface TVUIRLBandwidthMeter ()
@property (nonatomic, readwrite) uint64_t totalBytes;
@property (nonatomic, readwrite) uint64_t latestSpeed;
@property (nonatomic, assign) uint64_t previousTotalBytes;
@property (nonatomic, assign) uint64_t speedChangeRate;
@end

@implementation TVUIRLBandwidthMeter

- (instancetype)init {
    return [self initWithSpeedChangeRate:10];
}

- (instancetype)initWithSpeedChangeRate:(uint64_t)speedChangeRate {
    if (self = [super init]) {
        _speedChangeRate = speedChangeRate;
    }
    return self;
}

- (void)addBytesTransferred:(NSInteger)bytes {
    if (bytes > 0) self.totalBytes += (uint64_t)bytes;
}

- (TVUIRLBandwidthSnapshot)update {
    uint64_t speed = self.totalBytes - self.previousTotalBytes;
    self.latestSpeed = (self.speedChangeRate * speed + (100 - self.speedChangeRate) * self.latestSpeed) / 100;
    self.previousTotalBytes = self.totalBytes;
    TVUIRLBandwidthSnapshot snap = { .total = self.totalBytes, .speed = self.latestSpeed };
    return snap;
}

@end

File: DJIStreamDemo/RTMPServerC/tvu_irl_bandwidth_meter.c
/*
 * tvu_irl_bandwidth_meter.c
 */

#include "tvu_irl_bandwidth_meter.h"

tvu_irl_bandwidth_snapshot_t tvu_irl_bandwidth_meter_update(tvu_irl_bandwidth_meter_t *m) {
    uint64_t delta = m->total_bytes - m->previous_total_bytes;
    /* EMA-style 加权：k * 当前样本 + (100-k) * 历史 */
    m->latest_speed = (m->speed_change_rate * delta
                     + (100 - m->speed_change_rate) * m->latest_speed) / 100;
    m->previous_total_bytes = m->total_bytes;
    tvu_irl_bandwidth_snapshot_t s = { m->total_bytes, m->latest_speed };
    return s;
}
